### PR TITLE
[SpecDecode] phase2 performance tuning, compatible with scheduler overlap

### PR DESCRIPTION
**Pull request by pengchengneo**
image

## 4k input / 1k output benchserving logs and results

Raw logs/results bundle: https://gist.github.com/pengchengneo/328518a93ecd11366e5bea09f605d189

The gist contains:
- rank0 raw server logs for spec and no-spec
- benchserving `run.log` and `result.jsonl` for bsz 4/8/16/32/64
- parsed TTFT/ITL CSV
- parsed rank0 steady decode gen-throughput CSV

## Spec server launch command, rank0 shown

```bash
python3 -m sgl_jax.launch_server \
  --model-path /data/pc \
  --trust-remote-code \
  --tp-size 16 \
  --dp-size 4 \
  --ep-size 16 \
  --moe-backend epmoe \
  --nnodes 4 \
  --node-rank 0 \
  --dist-init-addr 10.16.2.6:30042 \
  --host 0.0.0.0 \
  --port 30271 \
  --page-size 256 \
  --context-length 262144 \
  --disable-radix-cache \
  --chunked-prefill-size 2048 \
  --dtype bfloat16 \
  --mem-fraction-static 0.84 \
  --swa-full-tokens-ratio 0.2 \
  --skip-server-warmup \
  --log-level info \
  --decode-log-interval 1 \
  --max-running-requests 128 \
  --dp-schedule-policy round_robin \
  --max-seq-len 4096 \
  --precompile-bs-paddings 1 4 8 16 32 64 128 \
  --precompile-token-paddings 4096 \
  --speculative-algorithm NEXTN \
  --speculative-num-steps 3 \
  --speculative-num-draft-tokens 4 \
  --speculative-eagle-topk 1 \
  --speculative-accept-threshold-single 1.0 \
  --speculative-accept-threshold-acc 1.0
```

## No-spec server launch command, rank0 shown

```bash
python3 -m sgl_jax.launch_server \
  --model-path /data/pc \
  --trust-remote-code \
  --tp-size 16 \
  --dp-size 4 \
  --ep-size 16 \
  --moe-backend epmoe \
  --nnodes 4 \
  --node-rank 0 \
  --dist-init-addr 10.16.2.6:30042 \
  --host 0.0.0.0 \
  --port 30271 \
  --page-size 256 \
  --context-length 262144 \
  --disable-radix-cache \
  --chunked-prefill-size 2048 \
  --dtype bfloat16 \
  --mem-fraction-static 0.84 \
  --swa-full-tokens-ratio 0.2 \
  --skip-server-warmup \
  --log-level info \
  --decode-log-interval 1 \
  --max-running-requests 128 \
  --dp-schedule-policy round_robin \
  --max-seq-len 4096 \
  --precompile-bs-paddings 1 4 8 16 32 64 128 \
  --precompile-token-paddings 4096
```

For ranks 1/2/3, only `--node-rank` and pod/log filename differ.

## Benchserving command template

```bash
for bs in 4 8 16 32 64; do
  num_prompts=$((bs * 3))
  python3 -m sgl_jax.bench_serving \
    --backend sgl-jax \
    --base-url http://127.0.0.1:30271 \
    --model /data/pc \
    --dataset-name random \
    --random-input-len 4096 \
    --random-output-len 1024 \
    --random-range-ratio 1 \
    --max-concurrency "$bs" \
    --num-prompts "$num_prompts" \
    --warmup-requests 0 \
    --output-file "$out/bs_${bs}/result.jsonl" \
    2>&1 | tee "$out/bs_${bs}/run.log"
done
```

## Benchserving results

| bsz | spec out tok/s | no-spec out tok/s | speedup | spec TTFT ms | no-spec TTFT ms | spec ITL ms | no-spec ITL ms |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 4 | 457.90 | 326.17 | 1.404x | 1865.45 | 2091.70 | 6.42 | 10.23 |
| 8 | 711.06 | 519.27 | 1.369x | 2173.81 | 2668.22 | 8.77 | 12.81 |
| 16 | 817.10 | 761.33 | 1.073x | 2691.87 | 3832.60 | 16.21 | 17.28 |
| 32 | 1168.70 | 1103.83 | 1.059x | 4072.27 | 6124.56 | 22.87 | 23.02 |
| 64 | 1162.40 | 1418.37 | 0.820x | 6989.83 | 10875.96 | 46.98 | 34.52 |

## Rank0 server-log steady decode gen throughput

| bsz | spec steady decode gen tok/s | no-spec steady decode gen tok/s | speedup |
| --- | --- | --- | --- |
| 4 | 1024.64 | 402.51 | 2.546x |
| 8 | 1681.14 | 669.43 | 2.511x |
| 16 | 2503.09 | 1046.65 | 2.392x |
| 32 | 4076.77 | 1711.67 | 2.382x |
| 64 | 5025.27 | 2629.48 | 1.911x |

Notes:
- TTFT/ITL are benchserving streaming-client metrics. TTFT includes scheduling and queueing under load, not only first-token compute.
- ITL is inter-stream-chunk latency for `sgl-jax`; it is not retokenized by accepted-token count for speculative decoding in this backend.

## Throughput comparison plot

Spec vs no-spec steady decode throughput comparison

### CODE REVIEW

**Comment by lingebeng** [comment]
cc @cjx0709 — CI is green now. Could you take another look?

**Comment by cjx0709** [comment]
image
this data seems strange, it looks different with the second sheet.

**Comment by pengchengneo** [comment]
> image this data seems strange, it looks different with the second sheet.

second picture is steady output throupht
this is bench-serving result,  and this is test for older commit, final result for 1c49097f8c73fd7a407128cd7a303978d22d0e5c put it here

The following test passes successfully:

```bash
cd test/srt && uv run python3 -m unittest test_speculative_decoding.TestSpeculativeDecoding.test_mmlu
```

The corresponding MMLU score is:

```text
0.797
```

This confirms that the CI test path is fixed and the speculative decoding behavior remains correct.

### 2. Qwen3-32B GSM8K Evaluation

Evaluated Qwen3-32B on GSM8K:

| Model     | Dataset | Metric   | Subset |  Num |  Score | Category |
| --------- | ------- | -------- | ------ | ---: | -----: | -------- |
| Qwen3-32B | gsm8k   | mean_acc | main   | 1319 | 0.9598 | default  |

The GSM8K result confirms that the fix does not introduce an accuracy regression for Qwen3-32B.

### 3. MiMo-V2-Flash AIME24 Sampling Validation

To further verify correctness for MiMo-V2-Flash, I ran AIME24 sampling validation with both greedy and non-greedy decoding settings:

| Decoding Setting                 | AIME24 Score | Correct |
| -------------------------------- | ------------ | ------- |
| greedy, temp=0                   | 0.8333       | 25/30   |
| non-greedy, temp=1.0, top_p=0.95 | 0.8667       | 26/30   |


These AIME24 results show that MiMo-V2-Flash produces stable and reasonable outputs under both deterministic and stochastic decoding settings.

## Throughput


**Server command**
```bash
LIBTPU_INIT_ARGS="--xla_tpu_dvfs_p_state=7" \
/jax-ai-image/sgl-jax/.venv/bin/python -m sgl_jax.launch_server \
  --model-path /models/MiMo-V2-Flash \
  --trust-remote-code \
  --enable-sequence-parallel \
  --tp-size 8 \
  --dp-size 2 \
  --ep-size 8 \
  --moe-backend fused_v2 \
  --nnodes 1 \
  --node-rank 0 \
  --host 0.0.0.0 \
  --port 30271 \
  --page-size 256 \
  --context-length 262144 \
  --disable-radix-cache \
  --chunked-prefill-size 2048 \
  --max-prefill-tokens 16384 \
  --dtype bfloat16 \
  --mem-fraction-static 0.84 \
  --swa-full-tokens-ratio 0.2 \
  --skip-server-warmup \
  --log-level info \
  --decode-log-interval 1 \
  --max-running-requests 256 \
  --dp-schedule-policy round_robin \
  --precompile-bs-paddings 1 4 8 16 32 64 128 256 \
  --precompile-token-paddings 4096 \
  --speculative-algorithm NEXTN \
  --speculative-num-steps 3 \
  --speculative-num-draft-tokens 4 \
  --speculative-eagle-topk 1 \
  --speculative-accept-threshold-single 1.0 \
  --speculative-accept-threshold-acc 1.0
```

**bench_serving command**
greedy：
```bash
for bs in 32 64 128; do
  num_prompts=$((bs * 3))
  /jax-ai-image/sgl-jax/.venv/bin/python -m sgl_jax.bench_serving \
    --backend sgl-jax \
    --base-url http://127.0.0.1:30271 \
    --model /models/MiMo-V2-Flash \
    --dataset-name random \
    --random-input-len 16384 \
    --random-output-len 4096 \
    --random-range-ratio 1 \
    --max-concurrency "$bs" \
    --num-prompts "$num_prompts" \
    --warmup-requests 0 \
    --output-file "/tmp/tpu_logs/bench-head4247-greedy-flash-16k-4k-bs${bs}-n${num_prompts}.jsonl"
done
```

non-greedy：
```bash
for bs in 32 64 128; do
  num_prompts=$((bs * 3))
  /jax-ai-image/sgl-jax/.venv/bin/python -m sgl_jax.bench_serving \
    --backend sgl-jax \
    --base-url http://127.0.0.1:30271 \
    --model /models/MiMo-V2-Flash \
    --dataset-name random \
    --random-input-len 16384 \
    --random-output-len 4096 \
    --random-range-ratio 1 \
    --max-concurrency "$bs" \
    --num-prompts "$num_prompts" \
    --warmup-requests 0 \
    --extra-request-body '{"sampling_params":{"temperature":0.1,"top_p":0.95,"max_new_tokens":4096,"ignore_eos":true}}' \
    --output-file "/tmp/tpu_logs/bench-head4247-nongreedy-flash-16k-4k-bs${bs}-n${num_prompts}.jsonl"
done
```

**Test Result**

| mode | bsz | num_prompts | Output tok/s | Peak output tok/s | Mean TTFT ms | P99 TTFT ms | Mean ITL ms | P99 ITL ms |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| greedy | 32 | 96 | 1702.00 | 4863.00 | 15403.42 | 42416.33 | 14.66 | 37.13 |
| greedy | 64 | 192 | 1840.19 | 5995.00 | 36130.70 | 82510.95 | 25.14 | 49.91 |
| greedy | 128 | 384 | 1979.53 | 7884.00 | 65984.69 | 165636.19 | 47.14 | 467.23 |
| non-greedy | 32 | 96 | 1370.44 | 4484.00 | 16739.38 | 43035.94 | 17.81 | 37.33 |
| non-greedy | 64 | 192 | 1727.81 | 5280.00 | 32821.64 | 83555.00 | 28.12 | 67.76 |
| non-greedy | 128 | 384 | 1824.04 | 6848.00 | 61619.19 | 165632.67 | 53.90 | 513.66 |

File: python/sgl_jax/srt/speculative/eagle_util.py [hunk 1]
```diff
@@ -642,19 +793,28 @@ def prepare_for_draft_decode(
...
     @classmethod
     def create_idle_input(
         cls,
         hidden_size: int,
         dtype: np.dtype,
         topk: int,
         capture_hidden_mode: CaptureHiddenMode,
+        num_steps: int = 1,
     ):
+        topk_shape = (0, num_steps, topk) if num_steps > 1 else (0, topk)
```
Comment (cjx0709) [inline python/sgl_jax/srt/speculative/eagle_util.py]: Now that this takes `num_steps` and branches the shape on it, the idle verify caller at line 1120 (`EagleVerifyInput.verify`, the `is_idle()` branch) still relies on the `num_steps=1` default — it doesn't pass anything. So for NEXTN with `speculative_num_steps > 1` the idle draft input gets `topk_index` shaped `(0, topk)` while the normal path produces `(0, num_steps, topk)`, i.e. a different rank. That caller has `self.spec_steps` available and should pass `num_steps=self.spec_steps`.

File: python/sgl_jax/srt/layers/attention/flashattention_backend.py [hunk 2]
```diff
@@ -212,16 +212,7 @@ def get_eagle_forward_metadata(self, batch: ModelWorkerBatch):
...
-            if batch.spec_info_padded.custom_mask.dtype == jnp.bool:
-                # FIXME(pc) rm this dtype convert
-                logger.warning(
-                    "batch.spec_info_padded.custom_mask type is  %s, it may make performance very low",
-                    batch.spec_info_padded.custom_mask.dtype,
-                )
-                metadata.custom_mask = batch.spec_info_padded.custom_mask.astype(jnp.int32)
-            else:
-                metadata.custom_mask = batch.spec_info_padded.custom_mask
+            metadata.custom_mask = batch.spec_info_padded.custom_mask
```
Comment (cjx0709) [inline python/sgl_jax/srt/layers/attention/flashattention_backend.py:215]: We dropped the old bool→int32 conversion + warning here, so the contract is now "caller must hand us a non-bool mask". If a bool mask ever sneaks in it'll silently corrupt things. Maybe worth a cheap assert that the dtype isn't bool so it fails loudly instead.

File: python/sgl_jax/srt/speculative/overlap_utils.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def publish_spec_decode_new_seq_lens(batch_output):
+    new_seq_lens = batch_output.next_draft_input.new_seq_lens
+    if new_seq_lens is not None and hasattr(new_seq_lens, "copy_to_host_async"):
+        new_seq_lens.copy_to_host_async()
+    return new_seq_lens
+
+
+def can_use_spec_decode_overlap(enable_overlap, spec_algorithm, batch) -> bool:
```
Comment (cjx0709) [inline python/sgl_jax/srt/speculative/overlap_utils.py:24]: This gate doesn't exclude grammar/vocab_mask/penalties, and the fused verify path doesn't handle those constraints. To be fair the non-fused `sample()` doesn't really apply them either — see the commented-out TODOs around eagle_util.py:1144 (custom sampler / penalty / grammar mask) — so the point isn't "fused regressed vs non-fused". It's that this PR routes more requests into the fused overlap path without the gate explicitly excluding sampling constraints the fused path currently can't honor, so they silently slip through. Worth excluding vocab_mask/grammar/penalty here.

Separately, the grammar lifecycle in the output-processor spec branch looks off: around line 517 it hands the whole accepted-token list to `req.grammar.accept_token(...)`, whereas the normal path advances the grammar one token at a time.

File: python/sgl_jax/srt/speculative/overlap_utils.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    if batch.forward_mode.is_mixed():
+        return False
+    if not batch.forward_mode.is_decode():
+        return False
+    if any(info.decoding_reqs for info in batch.reqs_info):
+        return False
+    return not (batch.return_logprob or batch.return_output_logprob_only)
+
+
+def can_use_spec_prefill_overlap(enable_overlap, spec_algorithm, batch) -> bool:
```
Comment (cjx0709) [inline python/sgl_jax/srt/speculative/overlap_utils.py:46]: Same kind of gap, but this one crashes instead of going silent. This gate only excludes logprob, not penalties/vocab_mask. The scheduler dispatches on it (`scheduler.py:2079`) with no try/except, and the entry (`base_worker.py:178`) raises `NotImplementedError` when `_can_use_fused_spec_prefill` is False — which is exactly the penalty / vocab_mask case. So a NEXTN topk=1 prefill with a penalty (pretty common) hits this gate and hard-crashes. Either align this with `_can_use_fused_spec_prefill` or fall back to `forward_batch_speculative_generation`.

File: python/sgl_jax/srt/speculative/draft_extend_fused.py [hunk 5]
```diff
@@ -0,0 +1,1953 @@
...
+            np.asarray(getattr(si, "top_ks", None)).reshape(-1)
+            if getattr(si, "top_ks", None) is not None
+            else None
+        )
+        _p = (
+            np.asarray(getattr(si, "top_ps", None), np.float32).reshape(-1)
+            if getattr(si, "top_ps", None) is not None
+            else None
+        )
+        _tv = float(_t[0]) if _t.size else 1.0
```
Comment (cjx0709) [inline python/sgl_jax/srt/speculative/draft_extend_fused.py]: `_tv/_kv/_pv` only read element 0 and then broadcast it across the whole batch, so a batch with mixed temperature/top_k/top_p ends up sampling everyone with request 0's params. And this is reachable — neither `can_use_spec_decode_overlap` nor `forward_batch_speculative_decode_overlap` requires `is_all_greedy` (unlike the non-overlap dispatch at `base_worker.py:265`), so a `temperature>0` batch lands here. Either build the per-request arrays from the full `temperatures`/`top_ks`/`top_ps`, or keep non-greedy batches out of the overlap path.

File: python/sgl_jax/srt/speculative/eagle_util.py (hunk 1 above)
Comment (lingebeng) [reply python/sgl_jax/srt/speculative/eagle_util.py]: I'll remove the create_idle_input func in https://github.com/sgl-project/sglang-jax/pull/1389

File: python/sgl_jax/srt/layers/attention/flashattention_backend.py (hunk 2 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/layers/attention/flashattention_backend.py:215]: add assert here @lingebeng

File: python/sgl_jax/srt/speculative/overlap_utils.py (hunk 3 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/speculative/overlap_utils.py:24]: currently , we have not compatible with vocab_mask/grammar/penalty both fused or non-fused, I think we can add TODO comments here, it will be supported by intern later @lingebeng

File: python/sgl_jax/srt/speculative/draft_extend_fused.py (hunk 5 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/speculative/draft_extend_fused.py]: please fix it @lingebeng

File: python/sgl_jax/srt/speculative/overlap_utils.py (hunk 4 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/speculative/overlap_utils.py:46]: add fallback logic here, or make it warning but no server crash @lingebeng

**Review by cjx0709 — APPROVED** [review]
LGTM
